hack/bundle-gen.py

```diff
@@ -129,26 +130,59 @@ def get_params():
 
     return args
 
+# Traverse through version directories to detect highest version
+def get_previous_version(work_dir, channel_name):
+    upstream_repo = "git@github.com:redhat-openshift-ecosystem/community-operators-prod.git"
+    upstream_branch = "main"
+    dir_name = "prev_version_determiner"
 
-# get_previous_version grabs the previous hive version (without the leading `v`) from
-# COMMUNITY_OPERATORS_HIVE_PKG_URL package yaml for the provided channel_name.
-def get_previous_version(channel_name):
-    http = urllib3.PoolManager()
-    r = http.request("GET", COMMUNITY_OPERATORS_HIVE_PKG_URL)
-    pkg_yaml = yaml.load(r.data.decode("utf-8"), Loader=yaml.FullLoader)
+    repo_full_path = os.path.join(work_dir, dir_name)
+    # clone git repo
+    try:
+        git.Repo.clone_from(upstream_repo, repo_full_path)
+    except:
+        print("Failed to clone repo {} to {}".format(upstream_repo, repo_full_path))
+        raise
+
+    # get to the right place on the filesystem
+    print("Working in %s" % repo_full_path)
+    os.chdir(repo_full_path)
+
+    repo = git.Repo(repo_full_path)
+
+    # Starting branch
+    print("Checkout latest {}".format(upstream_branch))
     try:
-        for channel in pkg_yaml["channels"]:
-            if channel["name"] == channel_name:
-                return channel["currentCSV"].replace("hive-operator.v", "")
+        repo.git.checkout(upstream_branch)
+    except:
+        print("Failed to checkout {}".format(upstream_branch))
+        raise
+
+    highest_version = "0.0.0"
+    try:
+        hive_dir = os.path.join(repo_full_path, HIVE_SUB_DIR)
+        for version in os.listdir(hive_dir):
+            if version == "ci.yaml":
+                continue
+            annotation_yaml_path = os.path.join(hive_dir, version, "metadata", "annotations.yaml")
+            with open(annotation_yaml_path, "r") as stream:
+                annotation_yaml = yaml.load(stream, Loader=yaml.SafeLoader)
+                version_channel = annotation_yaml["annotations"]["operators.operatorframework.io.bundle.channel.default.v1"]
+                if version_channel == channel_name:
+                    if semver.compare(version, highest_version) > 0:
+                        highest_version = version
     except:
         print(
             "Unable to determine previous hive version from {}",
-            COMMUNITY_OPERATORS_HIVE_PKG_URL,
+            upstream_repo,
         )
         raise
 
-    # Channel not found -- no previous version.
-    return None
+    if highest_version == "0.0.0":
+        # Channel not found -- no previous version.
+        return None
```

Comment:
I'm tempted to say this should be fatal. I mean, I guess we theoretically want this to work if we introduce a new channel... but I think the bigger risk is some bug resulting in not getting a prev_version, resulting in no `replaces`, and we don't notice. That'll mess up the upgrade graph, which is Bad™. I would rather have to hand-edit the CSV or temporarily comment out the error that one time we introduce a new channel.

(And please leave a comment explaining all of that.)